### PR TITLE
feat: added CUID type

### PR DESCRIPTION
**Pull request by therosbif**
### Summary of changes

Added a `cuid` type similar to the current `uuid` as discussed in #419.

I'm currently not using the `isCuid` function from `@paralleldrive/cuid2` for deserialization because I'm not sure how the importing would work in the string templates.
I'm making the same checks manually minus the string check

```ts
// @paralleldrive/cuid2

const isCuid = (id, { minLength = 2, maxLength = bigLength } = {}) => {
  const length = id.length;
  const regex = /^[0-9a-z]+$/;

  try {
    if (
      typeof id === "string" &&
      length >= minLength &&
      length <= maxLength &&
      regex.test(id)
    )
      return true;
  } finally {
  }

  return false;
};
```

### Relinquishment of Rights

Please mark following checkbox to confirm that you relinquish all rights of your changes:

- [x] I waive and relinquish all rights regarding this changes (including code, text, and images) to Deepkit UG (limited), Germany. This changes (including code, text, and images) are under MIT license without name attribution, copyright notice, and permission notice requirement.

### CODE REVIEW

File: packages/type/src/utils.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@
  */
 
 import { stringify, v4 } from 'uuid';
+import { init } from '@paralleldrive/cuid2';
```
Comment (marcj) [inline packages/type/src/utils.ts:12]: Is this a big dependency? Can we move it to optional dependency, so it doesn't come automatically with deepkit/type?

**Comment by marcj** [comment]
Cool, thank you! A few tests would be great :)

File: packages/type/src/utils.ts (hunk 1 above)
Comment (therosbif) [reply packages/type/src/utils.ts:12]: It's only 127 lines: https://github.com/paralleldrive/cuid2/blob/main/src/index.js
but it depends on @noble/hashes.
It's total install size is 750kb: https://packagephobia.com/result?p=%40paralleldrive%2Fcuid2

I can move it to peer deps if you prefer

**Comment by therosbif** [comment]
Something seems to be wrong with the linking step that's stopping me from running the tests:
```
Cannot find module '@deepkit/type' from 'node_modules/@deepkit/rpc/dist/cjs/src/model.js'
```
Do you know what could be causing this? I've run all the steps in DEVELOPMENT.md successfully.

**Comment by therosbif** [comment]
I'm getting an error on the  'optional basics' and 'nullable basics' tests in `packages/type/tests/type-spec.spec.ts` that I can't figure out how to fix:
```
● optional basics

    TypeError: Cannot read properties of undefined (reading 'length')

      at self (eval at build (../core/src/compiler.ts:95:20), <anonymous>:17:15)
      at self (eval at build (../core/src/compiler.ts:95:20), <anonymous>:19:32)
      at serialize (src/serializer-facade.ts:172:12)
      at roundTrip (tests/type-spec.spec.ts:15:27)
      at Object.<anonymous> (tests/type-spec.spec.ts:473:12)

  ● nullable basics

    TypeError: Cannot read properties of undefined (reading 'length')

      at self (eval at build (../core/src/compiler.ts:95:20), <anonymous>:17:15)
      at self (eval at build (../core/src/compiler.ts:95:20), <anonymous>:19:32)
      at serialize (src/serializer-facade.ts:172:12)
      at roundTrip (tests/type-spec.spec.ts:15:27)
      at Object.<anonymous> (tests/type-spec.spec.ts:539:12)
```

Where can I set the nullable behaviour of a new type? (If that's actually the issue)

**Pull request closed without merging**